File: src/runtime/cpp/std/argparse.cpp
#include "argparse.h"

#include "built_in/contains.h"
#include "built_in/string_ops_fwd.h"
#include "sys.h"
#include "core/exceptions.h"

Namespace::Namespace()
    : values(rc_dict_from_value(dict<str, object>{})) {}

Namespace::Namespace(const Object<dict<str, object>>& values_in)
    : values(values_in) {}

_ArgSpec::_ArgSpec()
    : names(rc_list_from_value(list<str>{})),
      action(str("")),
      choices(rc_list_from_value(list<str>{})),
      default_value(object()),
      help_text(str("")),
      is_optional(false),
      dest(str("")) {}

_ArgSpec::_ArgSpec(
    const Object<list<str>>& names_in,
    const str& action_in,
    const Object<list<str>>& choices_in,
    const object& default_value_in,
    const str& help_text_in
)
    : names(names_in),
      action(action_in),
      choices(choices_in),
      default_value(default_value_in),
      help_text(help_text_in),
      is_optional(false),
      dest(str("")) {
    is_optional = py_len(names) > int64(0) && py_startswith(py_list_at_ref(names, int64(0)), str("-"));
    if (is_optional) {
        str base = py_replace(py_lstrip_chars(py_list_at_ref(names, py_len(names) - int64(1)), str("-")), str("-"), str("_"));
        dest = base;
    } else if (py_len(names) > int64(0)) {
        dest = py_list_at_ref(names, int64(0));
    }
}

ArgumentParser::ArgumentParser(const str& description_in)
    : description(description_in),
      _specs(rc_list_from_value(list<_ArgSpec>{})) {}

void ArgumentParser::add_argument(
    const str& name0,
    const str& name1,
    const str& name2,
    const str& name3,
    const str& help,
    const str& action,
    const Object<list<str>>& choices,
    const object& default_value
) {
    Object<list<str>> names = rc_list_from_value(list<str>{});
    if (name0 != str("")) py_list_append_mut(names, name0);
    if (name1 != str("")) py_list_append_mut(names, name1);
    if (name2 != str("")) py_list_append_mut(names, name2);
    if (name3 != str("")) py_list_append_mut(names, name3);
    if (py_len(names) == int64(0)) {
        throw ::std::runtime_error("add_argument requires at least one name");
    }
    py_list_append_mut(_specs, _ArgSpec(names, action, choices, default_value, help));
}

void ArgumentParser::_fail(const str& msg) const {
    if (msg != str("")) {
        write_stderr(str("error: ") + msg + str("\n"));
    }
    throw SystemExit(2);
}

Object<dict<str, object>> ArgumentParser::parse_args(const ::std::optional<Object<list<str>>>& argv_in) const {
    Object<list<str>> args = rc_list_from_value(list<str>{});
    if (argv_in == ::std::nullopt) {
        args = rc_list_from_value(py_list_slice_copy(::argv, int64(1), py_len(::argv)));
    } else {
        args = *argv_in;
    }

    Object<list<_ArgSpec>> specs_pos = rc_list_from_value(list<_ArgSpec>{});
    Object<list<_ArgSpec>> specs_opt = rc_list_from_value(list<_ArgSpec>{});
    for (const auto& spec : _specs) {
        if (spec.is_optional) {
            py_list_append_mut(specs_opt, spec);
        } else {
            py_list_append_mut(specs_pos, spec);
        }
    }

    Object<dict<str, int64>> by_name = rc_dict_from_value(dict<str, int64>{});
    int64 spec_i = int64(0);
    for (const auto& spec : specs_opt) {
        for (const auto& name : spec.names) {
            (*by_name)[name] = spec_i;
        }
        spec_i += int64(1);
    }

    Object<dict<str, object>> values = rc_dict_from_value(dict<str, object>{});
    for (const auto& spec : _specs) {
        if (spec.action == str("store_true")) {
            if (py_runtime_value_exact_is<bool>(spec.default_value)) {
                (*values)[spec.dest] = spec.default_value;
            } else {
                (*values)[spec.dest] = object(false);
            }
        } else if (!py_is_none(spec.default_value)) {
            (*values)[spec.dest] = spec.default_value;
        } else {
            (*values)[spec.dest] = object();
        }
    }

    int64 pos_i = int64(0);
    int64 i = int64(0);
    while (i < py_len(args)) {
        str tok = py_list_at_ref(args, i);
        if (py_startswith(tok, str("-"))) {
            if (!py_contains(by_name, tok)) {
                _fail(str("unknown option: ") + tok);
            }
            const _ArgSpec& spec = py_list_at_ref(specs_opt, py_at(by_name, tok));
            if (spec.action == str("store_true")) {
                (*values)[spec.dest] = object(true);
                i += int64(1);
                continue;
            }
            if (i + int64(1) >= py_len(args)) {
                _fail(str("missing value for option: ") + tok);
            }
            str val = py_list_at_ref(args, i + int64(1));
            if (py_len(spec.choices) > int64(0) && !py_contains(spec.choices, val)) {
                _fail(str("invalid choice for ") + tok + str(": ") + val);
            }
            (*values)[spec.dest] = object(val);
            i += int64(2);
            continue;
        }

        if (pos_i >= py_len(specs_pos)) {
            _fail(str("unexpected extra argument: ") + tok);
        }
        const _ArgSpec& spec = py_list_at_ref(specs_pos, pos_i);
        (*values)[spec.dest] = object(tok);
        pos_i += int64(1);
        i += int64(1);
    }

    if (pos_i < py_len(specs_pos)) {
        _fail(str("missing required argument: ") + py_list_at_ref(specs_pos, pos_i).dest);
    }

    return values;
}

File: src/runtime/dart/std/argparse.dart
import '../built_in/py_runtime.dart';
import '../std/sys.dart' as sys;

class Namespace {
  dynamic values;
  Namespace([dynamic values = null]) : values = values ?? {};
}

class _ArgSpec {
  late List<String> names;
  late String action;
  late List<String> choices;
  late dynamic default_;
  late String helpText;
  late bool isOptional;
  late String dest;

  _ArgSpec(
    dynamic names, [
    String action = "",
    dynamic choices = const [],
    dynamic default_ = null,
    String helpText = "",
  ]) {
    this.names = List<String>.from(names);
    this.action = action;
    this.choices = List<String>.from(choices);
    this.default_ = default_;
    this.helpText = helpText;
    isOptional = this.names.isNotEmpty && this.names.first.startsWith("-");
    if (isOptional) {
      dest = this.names.last.replaceFirst(RegExp(r"^-+"), "").replaceAll("-", "_");
    } else {
      dest = this.names.first;
    }
  }
}

class ArgumentParser {
  late String description;
  late List<_ArgSpec> _specs;

  ArgumentParser([String description = ""]) {
    this.description = description;
    _specs = <_ArgSpec>[];
  }

  void add_argument(
    String name0, [
    dynamic name1 = "",
    dynamic name2 = "",
    dynamic name3 = "",
    String help = "",
    String action = "",
    dynamic choices = const [],
    dynamic default_ = null,
  ]) {
    if (action == "" && help == "" && name1 is String && name1 == "store_true") {
      action = "store_true";
      name1 = "";
    }
    if (action == "" && help == "" && choices is List && choices.isEmpty && default_ == null && name2 is List) {
      choices = name2;
      name2 = "";
      if (name3 is String && name3 != "") {
        default_ = name3;
        name3 = "";
      }
    }

    final names = <String>[];
    if (name0 != "") names.add(name0);
    if (name1 is String && name1 != "") names.add(name1);
    if (name2 is String && name2 != "") names.add(name2);
    if (name3 is String && name3 != "") names.add(name3);
    if (names.isEmpty) {
      throw ValueError("add_argument requires at least one name");
    }
    _specs.add(_ArgSpec(names, action, choices, default_, help));
  }

  Never _fail(String msg) {
    if (msg != "") {
      sys.write_stderr("error: $msg\n");
    }
    throw SystemExit(2);
  }

  dynamic parse_args([dynamic argv = null]) {
    final args = argv == null ? List<dynamic>.from(pytraSlice(sys.argv, 1, null)) : List<dynamic>.from(argv);
    final specsPos = <_ArgSpec>[];
    final specsOpt = <_ArgSpec>[];
    for (final s in _specs) {
      if (s.isOptional) {
        specsOpt.add(s);
      } else {
        specsPos.add(s);
      }
    }

    final byName = <String, int>{};
    for (var i = 0; i < specsOpt.length; i++) {
      for (final name in specsOpt[i].names) {
        byName[name] = i;
      }
    }

    final values = <String, dynamic>{};
    for (final s in _specs) {
      if (s.action == "store_true") {
        values[s.dest] = s.default_ is bool ? s.default_ : false;
      } else {
        values[s.dest] = s.default_;
      }
    }

    var posI = 0;
    var i = 0;
    while (i < args.length) {
      final tok = pytraStr(args[i]);
      if (tok.startsWith("-")) {
        if (!byName.containsKey(tok)) {
          _fail("unknown option: $tok");
        }
        final spec = specsOpt[byName[tok] as int];
        if (spec.action == "store_true") {
          values[spec.dest] = true;
          i += 1;
          continue;
        }
        if (i + 1 >= args.length) {
          _fail("missing value for option: $tok");
        }
        final val = pytraStr(args[i + 1]);
        if (spec.choices.isNotEmpty && !pytraContains(spec.choices, val)) {
          _fail("invalid choice for $tok: $val");
        }
        values[spec.dest] = val;
        i += 2;
        continue;
      }

      if (posI >= specsPos.length) {
        _fail("unexpected extra argument: $tok");
      }
      final spec = specsPos[posI];
      values[spec.dest] = tok;
      posI += 1;
      i += 1;
    }

    if (posI < specsPos.length) {
      _fail("missing required argument: ${specsPos[posI].dest}");
    }
    return values;
  }
}
